### PR TITLE
Fix Assembly tests for netfx

### PR DESCRIPTION
**Pull request by hughbe**
Most of these are expected due to differences in the implementation of Assembly in the runtimes

Not sure about the `GetModule(module.Name)` tests, though.

Contributes to #17723

### CODE REVIEW

**Comment by danmosemsft** [comment]
I'm also not sure about `GetModule` either and it's not obvious from code (managed side seems same as desktop) @rahku should answer.

File: src/System.Runtime/tests/System/Reflection/AssemblyTests.cs [hunk 1]
```diff
@@ -257,55 +265,99 @@ public static void GetManifestResourceStream()
...
         [Fact]
         public static void Test_LoadFile()
         {
             var assem = typeof(AssemblyTests).Assembly;
             string path = "System.Runtime.Tests.dll";
             string fullpath = Path.GetFullPath(path);
-            Assert.Throws<ArgumentNullException>("path", () => Assembly.LoadFile(null));
-            Assert.Throws<ArgumentException>(() => Assembly.LoadFile(path));
             var loadfile1 = Assembly.LoadFile(fullpath);
-            Assert.NotEqual(assem, loadfile1);
```
Comment (rahku) [inline src/System.Runtime/tests/System/Reflection/AssemblyTests.cs]: This is an important assert. Why did you remove this. I believe it should pass for both .net core & desktop

File: src/System.Runtime/tests/System/Reflection/AssemblyTests.cs [hunk 2]
```diff
@@ -257,55 +265,99 @@ public static void GetManifestResourceStream()
...
+            Assert.Equal(assem.FullName, loadfile1.FullName);
             string dir = Path.GetDirectoryName(fullpath);
             fullpath = Path.Combine(dir, ".", path);
             var loadfile2 = Assembly.LoadFile(fullpath);
-            Assert.Equal(loadfile1,loadfile2);
```
Comment (rahku) [inline src/System.Runtime/tests/System/Reflection/AssemblyTests.cs]: this as well ..we want to compare assembly object ...not just display name

File: src/System.Runtime/tests/System/Reflection/AssemblyTests.cs [hunk 3]
```diff
@@ -197,8 +197,16 @@ public static void GetModules_GetModule(Assembly assembly)
             Assert.NotEmpty(assembly.GetModules());
             foreach (Module module in assembly.GetModules())
             {
-                Assert.NotNull(module);
-                Assert.Equal(module, assembly.GetModule(module.Name));
```
Comment (rahku) [inline src/System.Runtime/tests/System/Reflection/AssemblyTests.cs]: in my experiment i see the same behavior on desktop & .net core... which is 
            Assert.NotNull(module);
            Assert.Equal(module, assembly.GetModule(module.Name));

Comment (hughbe) [inline src/System.Runtime/tests/System/Reflection/AssemblyTests.cs]: Hmm, I get the following test failures: `msbuild /T:BuildAndTest /P:TargetGroup=netfx`

```
      System.Reflection.Tests.AssemblyTests.GetModules_GetModule(assembly: mscorlib, Version=4.0.0.0, Culture=neutral,
  PublicKeyToken=b77a5c561934e089) [FAIL]
        Assert.Equal() Failure
        Expected: CommonLanguageRuntimeLibrary
        Actual:   (null)
        Stack Trace:
          C:\Users\hughb\Documents\GitHub\corefx\src\System.Runtime\tests\System\Reflection\AssemblyTests.cs(200,0): at
   System.Reflection.Tests.AssemblyTests.GetModules_GetModule(Assembly assembly)
      System.Reflection.Tests.AssemblyTests.GetModules_GetModule(assembly: mscorlib, Version=4.0.0.0, Culture=neutral,
  PublicKeyToken=b77a5c561934e089) [FAIL]
        Assert.Equal() Failure
        Expected: CommonLanguageRuntimeLibrary
        Actual:   (null)
        Stack Trace:
          C:\Users\hughb\Documents\GitHub\corefx\src\System.Runtime\tests\System\Reflection\AssemblyTests.cs(200,0): at
   System.Reflection.Tests.AssemblyTests.GetModules_GetModule(Assembly assembly)
```

File: src/System.Runtime/tests/System/Reflection/AssemblyTests.cs (hunk 1 above)
Comment (hughbe) [inline src/System.Runtime/tests/System/Reflection/AssemblyTests.cs]: Hmm, I get the following test failures: `msbuild /T:BuildAndTest /P:TargetGroup=netfx`

```
        Assert.NotEqual() Failure
        Expected: Not System.Runtime.Tests, Version=4.2.0.0, Culture=neutral, PublicKeyToken=9d77cc7ad39b68eb
        Actual:   System.Runtime.Tests, Version=4.2.0.0, Culture=neutral, PublicKeyToken=9d77cc7ad39b68eb
        Stack Trace:
          C:\Users\hughb\Documents\GitHub\corefx\src\System.Runtime\tests\System\Reflection\AssemblyTests.cs(303,0): at
   System.Reflection.Tests.AssemblyTests.Test_LoadFile()
```

File: src/System.Runtime/tests/System/Reflection/AssemblyTests.cs (hunk 2 above)
Comment (hughbe) [inline src/System.Runtime/tests/System/Reflection/AssemblyTests.cs]: Hmm, I get the following test failures: `msbuild /T:BuildAndTest /P:TargetGroup=netfx`

```
      System.Reflection.Tests.AssemblyTests.Test_LoadFile [FAIL]
        Assert.Equal() Failure
        Expected: System.Runtime.Tests, Version=4.2.0.0, Culture=neutral, PublicKeyToken=9d77cc7ad39b68eb
        Actual:   System.Runtime.Tests, Version=4.2.0.0, Culture=neutral, PublicKeyToken=9d77cc7ad39b68eb
        Stack Trace:
          C:\Users\hughb\Documents\GitHub\corefx\src\System.Runtime\tests\System\Reflection\AssemblyTests.cs(307,0): at
   System.Reflection.Tests.AssemblyTests.Test_LoadFile()
``

**Comment by danmosemsft** [comment]
@rahku can you help @hughbe with his isues above?

File: src/System.Runtime/tests/System/Reflection/AssemblyTests.cs (hunk 1 above)
Comment (rahku) [inline src/System.Runtime/tests/System/Reflection/AssemblyTests.cs]: yes i am getting the same behavior. You should keep the assert for .net core same as what we had before and for desktop the assert should be Assert.Equal(assem, loadfile1);
It is important to compare the assembly and not just the display name

File: src/System.Runtime/tests/System/Reflection/AssemblyTests.cs (hunk 2 above)
Comment (rahku) [inline src/System.Runtime/tests/System/Reflection/AssemblyTests.cs]: same here as well...we need to compare assembly and jot just display name

File: src/System.Runtime/tests/System/Reflection/AssemblyTests.cs (hunk 3 above)
Comment (rahku) [inline src/System.Runtime/tests/System/Reflection/AssemblyTests.cs]: In this case we are performing test for type containing List<T> class. This is in System.Collections.dll for .net core and in mscorlib.dll for desktop. For single module assembly typically module name is same is file name. But for mscorlib.dll it is specifically specified as CommonLanguageRuntimeLibrary in IL. module.Name property returns the name of the module with the path removed. However in this test we want to find the name of module and not path name to provide as input to assembly.GetModule(). So using module.Name to get that is wrong. module.ToString is the correct api which provides the name of module. So you should just replace all instances of
module.Name
with 
module.ToString()
in this file

**Comment by hughbe** [comment]
Sorry for the delay, I've addressed the PR feedback.

**Comment by hughbe** [comment]
Test Innerloop Windows_NT Release x64 Build and Test

File: src/System.Runtime/tests/System/Reflection/AssemblyTests.cs [hunk 4]
```diff
@@ -261,55 +260,116 @@ public static void GetManifestResourceStream()
...
+                Assert.NotEqual(currentAssembly, loadedAssembly1);
+            }
+
+            string dir = Path.GetDirectoryName(fullRuntimeTestsPath);
+            fullRuntimeTestsPath = Path.Combine(dir, ".", RuntimeTestsDll);
+
+            Assembly loadedAssembly2 = Assembly.LoadFile(fullRuntimeTestsPath);
+            if (PlatformDetection.IsFullFramework)
+            {
+                Assert.NotEqual(loadedAssembly1, loadedAssembly2);
```
Comment (rahku) [inline src/System.Runtime/tests/System/Reflection/AssemblyTests.cs]: This is interesting to know that adding "." in the path causes Assembly.LoadFile to load again on desktop.

**Comment by rahku** [comment]
LGTM

**Comment by rahku** [comment]
Thanks for the contribution.

File: src/System.Runtime/tests/System/Reflection/AssemblyTests.cs (hunk 4 above)
Comment (hughbe) [inline src/System.Runtime/tests/System/Reflection/AssemblyTests.cs]: Is that behaviour something we want to align with in .NET Core?

Comment (rahku) [inline src/System.Runtime/tests/System/Reflection/AssemblyTests.cs]: No I don't think so. Behavior of Assembly.Load* apis in .net core are very different from desktop and that is by design.

Comment (rahku) [inline src/System.Runtime/tests/System/Reflection/AssemblyTests.cs]: in this particular case behavior of .net core seems more appropiate
